<a href="https://colab.research.google.com/github/williamokomba/Molinga-school-projects/blob/master/Moringa_Data_Science_Prep_W3_Independent_Project_2020_09_william_okomba_Python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##week 3 IP


In [232]:
#WHAT TO WORK ON
#Recall the basics of Python programming for data science. 
#Write code and document your workflow in a programming environment.
#Understand mechanisms for missing data, outliers and analytic implications.

In [233]:
#you will be working as Data Scientist for MTN Cote d'Ivoire, a leading telecom company and you will be solving for the following research question.

#Currently MTN Cote d'Ivoire would like to upgrade its technology infrastructure for its mobile users in Ivory Coast. Studying the given dataset, 
#how does MTN Cote d'Ivoire go about the upgrade of its infrastructure strategy within the given cities?
#Your final deliverable will be a Data Report which will comprise the following sections
#
#Business Understanding ,Data Understanding ,Data Preparation 
#Analysis ,Recommendation ,Evaluation

In [234]:
#Below are some questions that can get you started;

#1.Which ones were the most used city for the three days?
#2. Which cities were the most used during business and home hours?
#3. Most used city for the three days?
#etc. 

In [235]:
#importing pandas as pd
#
import pandas as pd

##loading the datasets

In [236]:
#i.load geographical decription
#
df_cells_geo_description= pd.read_excel("cells_geo_description.xlsx")
df_cells_geo_description.head()

Column name        Description  Format
0        VILLES               City  String
1        STATUS  In Service or not  String
2  LOCALISATION  in ABIDJAN or not  String
3    DECOUPZONE  Geographical Zone  String
4      ZONENAME       Name of Zone  String

In [237]:
#ii.load cdr description
#
df_CDR_description= pd.read_excel("CDR_description.xlsx")
df_CDR_description.head()

Column name                                        Description   Format
0          PRODUCT                                       Voice or SMS   String
1            VALUE                                      Billing price  Integer
2        DATE_TIME               Time in format yyyy-MM-dd hh:mm:ss.0   String
3     CELL_ON_SITE  Which cell in the site was used (not needed here)  Integer
4  DW_A_NUMBER_INT  Anonymized phone number of the person for whic...   String

In [238]:
#1a.load cell geo dataset
# 
df_cells_geo= pd.read_csv("cells_geo.csv", header= 0, index_col= 0, delimiter=";")
df_cells_geo

VILLES      STATUS LOCALISATION  ...          AREA     CELL_ID   SITE_CODE
0       ADJAME  In Service      ABIDJAN  ...           NaN  fbb2e75af3  0bb492ef17
1       ADJAME  In Service      ABIDJAN  ...           NaN  1502501d78  0bb492ef17
2       ADJAME  In Service      ABIDJAN  ...           NaN  0c2e182a0c  0bb492ef17
3      ASSINIE  In Service    INTERIEUR  ...           NaN  69bdfc69fd  8a1e6ce31f
4      ASSINIE  In Service    INTERIEUR  ...           NaN  7a1029d2ab  8a1e6ce31f
...        ...         ...          ...  ...           ...         ...         ...
3969  ZUENOULA  In Service    INTERIEUR  ...  CENTRE NORTH  1cfc2d7ab1  f91ed8d2e9
3970  ZUENOULA  In Service    INTERIEUR  ...  CENTRE NORTH  3ae1708153  f91ed8d2e9
3971  ZUENOULA  In Service    INTERIEUR  ...  CENTRE NORTH  cb71f8a29c  61bee9c70c
3972  ZUENOULA  In Service    INTERIEUR  ...  CENTRE NORTH  d4a6a0b645  61bee9c70c
3973  ZUENOULA  In Service    INTERIEUR  ...  CENTRE NORTH  f9f3e42589  61bee9c70c

[3974 rows x 11 columns]

In cells_geo.cs dataset, we had 3974 cells.



In [239]:
#1b.check duplicates based on cell id and dropping them.
#
df_cell_geo_1= df_cells_geo.drop_duplicates(['CELL_ID'], keep='last')
df_cell_geo_1

VILLES      STATUS LOCALISATION  ...          AREA     CELL_ID   SITE_CODE
0       ADJAME  In Service      ABIDJAN  ...           NaN  fbb2e75af3  0bb492ef17
1       ADJAME  In Service      ABIDJAN  ...           NaN  1502501d78  0bb492ef17
2       ADJAME  In Service      ABIDJAN  ...           NaN  0c2e182a0c  0bb492ef17
3      ASSINIE  In Service    INTERIEUR  ...           NaN  69bdfc69fd  8a1e6ce31f
4      ASSINIE  In Service    INTERIEUR  ...           NaN  7a1029d2ab  8a1e6ce31f
...        ...         ...          ...  ...           ...         ...         ...
3969  ZUENOULA  In Service    INTERIEUR  ...  CENTRE NORTH  1cfc2d7ab1  f91ed8d2e9
3970  ZUENOULA  In Service    INTERIEUR  ...  CENTRE NORTH  3ae1708153  f91ed8d2e9
3971  ZUENOULA  In Service    INTERIEUR  ...  CENTRE NORTH  cb71f8a29c  61bee9c70c
3972  ZUENOULA  In Service    INTERIEUR  ...  CENTRE NORTH  d4a6a0b645  61bee9c70c
3973  ZUENOULA  In Service    INTERIEUR  ...  CENTRE NORTH  f9f3e42589  61bee9c70c

[3916 rows x 11 columns]

The In cells_geo.cs dataset, has 58 dublicate values based on cell_Id.

duplicates = (total cells- cells number after duplicates are dropped).

duplicate = 3974 - 3916.

duplicate = 58

In [241]:
#1c.check missing values from differernt columns from cell_geo Dataset colunm by colunm.
#
df_cells_geo.AREA.unique()


array([nan, 'ABIDJAN NORTH', 'NORTH WEST', 'EAST', 'ABIDJAN SOUTH',
       'SOUTH WEST', 'CENTRE NORTH'], dtype=object)

Three columns had missing data;status,zone name, and area

There were 56 missing values from the status,zone name, and area.

In [244]:
#1d. check missing based on cell id and dropping them, get the final clean dataset.
#
df_cell_geo_2=df_cell_geo_1[~df_cell_geo_1['CELL_ID'].isin(df_missing_values['CELL_ID'])]
df_cell_geo_2


VILLES      STATUS LOCALISATION  ...           AREA     CELL_ID   SITE_CODE
23       ABOBO  In Service      ABIDJAN  ...  ABIDJAN NORTH  f740406877  0f34c36ac5
24       ABOBO  In Service      ABIDJAN  ...  ABIDJAN NORTH  959e54c8f7  0f34c36ac5
25       ABOBO  In Service      ABIDJAN  ...  ABIDJAN NORTH  56b7e7000c  0f34c36ac5
26       ABOBO  In Service      ABIDJAN  ...  ABIDJAN NORTH  bc51c17a6f  0f34c36ac5
27       ABOBO  In Service      ABIDJAN  ...  ABIDJAN NORTH  e460975c9f  0f34c36ac5
...        ...         ...          ...  ...            ...         ...         ...
3969  ZUENOULA  In Service    INTERIEUR  ...   CENTRE NORTH  1cfc2d7ab1  f91ed8d2e9
3970  ZUENOULA  In Service    INTERIEUR  ...   CENTRE NORTH  3ae1708153  f91ed8d2e9
3971  ZUENOULA  In Service    INTERIEUR  ...   CENTRE NORTH  cb71f8a29c  61bee9c70c
3972  ZUENOULA  In Service    INTERIEUR  ...   CENTRE NORTH  d4a6a0b645  61bee9c70c
3973  ZUENOULA  In Service    INTERIEUR  ...   CENTRE NORTH  f9f3e42589  61bee9c70c

[3860 rows x 11 columns]



clean dataset has 3860 cells.

missing dataset = df_cells_geo_1 - df_cells_geo_2

missing data = 3916- 3860 = 56.



In [265]:
#1di. using dropna function to drop missing values
df_cells = df_cell_geo_1.dropna(axis=1)
df_cells

VILLES LOCALISATION    DECOUPZONE  ...     REGION     CELL_ID   SITE_CODE
0       ADJAME      ABIDJAN  "Abidjan_EST  ...     LAGUNE  fbb2e75af3  0bb492ef17
1       ADJAME      ABIDJAN  "Abidjan_EST  ...     LAGUNE  1502501d78  0bb492ef17
2       ADJAME      ABIDJAN  "Abidjan_EST  ...     LAGUNE  0c2e182a0c  0bb492ef17
3      ASSINIE    INTERIEUR     Grand-EST  ...  SUD COMOE  69bdfc69fd  8a1e6ce31f
4      ASSINIE    INTERIEUR     Grand-EST  ...  SUD COMOE  7a1029d2ab  8a1e6ce31f
...        ...          ...           ...  ...        ...         ...         ...
3969  ZUENOULA    INTERIEUR    Grand-NORD  ...   MARAHOUE  1cfc2d7ab1  f91ed8d2e9
3970  ZUENOULA    INTERIEUR    Grand-NORD  ...   MARAHOUE  3ae1708153  f91ed8d2e9
3971  ZUENOULA    INTERIEUR    Grand-NORD  ...   MARAHOUE  cb71f8a29c  61bee9c70c
3972  ZUENOULA    INTERIEUR    Grand-NORD  ...   MARAHOUE  d4a6a0b645  61bee9c70c
3973  ZUENOULA    INTERIEUR    Grand-NORD  ...   MARAHOUE  f9f3e42589  61bee9c70c

[3916 rows x 8 columns]

using dropna function drops some columns thus i used isin function.

In [245]:
#1e.descriptive summary
#
df_location = df_cell_geo_2['CELL_ID'].groupby(df_cell_geo_2['LOCALISATION']).describe()
df_location

count unique         top freq
LOCALISATION                              
ABIDJAN       1605   1605  30a5636053    1
INTERIEUR     2255   2255  1e6c1bbd5a    1

In [246]:
#1f. getting the percentage propotion
#
interieur = 2255/3860*100
interieur

58.41968911917098

In [ ]:
abidjan = 1605/3860*100
abidjan

this means interiers control 58.4% of the calls while abidjan has 41.6% if the two are compared based on localisation.

In [247]:
#1g.checking cleaned  dataset.
#
df_location_area = df_cell_geo_1.groupby(df_cell_geo_2['AREA']).size()
df_location_area

AREA
ABIDJAN NORTH    1181
ABIDJAN SOUTH     642
CENTRE NORTH      610
EAST              334
NORTH WEST        562
SOUTH WEST        531
dtype: int64

this is the summary of the clean dataset.

In [248]:
#2a.load dataset1 
#
df_dataset1= pd.read_csv("dataset1.csv")
df_dataset1

PRODUTC  VALUE               DATETIME  ... COUNTRY_B     CELL_ID     SITE_ID
0      Voice      0  2012-05-06 23:04:37.0  ...     19e2e  /TJNe+Mmtu  /TJNe+Mmtu
1        sms      0  2012-05-06 23:05:25.0  ...     19e2e  /+cKZKIp41  /+cKZKIp41
2        sms     25  2012-05-06 23:05:41.0  ...     19e2e  /u0FSD+ahi  /u0FSD+ahi
3      Voice     56  2012-05-06 23:05:42.0  ...     19e2e  +cNeJzsTp3  +cNeJzsTp3
4      Voice     14  2012-05-06 23:06:29.0  ...     19e2e  051KhYDCpv  051KhYDCpv
...      ...    ...                    ...  ...       ...         ...         ...
4996     sms    100  2012-05-07 00:01:54.0  ...     OTHER  0c2371d9bc  1e751d08a4
4997   Voice      1  2012-05-07 00:01:54.0  ...     19e2e  5fb791cf56  65efe2bc62
4998   Voice      0  2012-05-07 00:01:54.0  ...     19e2e  7773627c12  9ef9616350
4999   Voice      0  2012-05-07 00:01:54.0  ...     19e2e  3f4af49d6c  969b8d0e39
5000   Voice      7  2012-05-07 00:01:54.0  ...     19e2e  6f4d9b77a5  ebfc58af70

[5001 rows x 10 columns]

In dataset1 dataset, we had 5001 cells.

In [249]:
#2b.rename colunm for dataset 1
#
df_dataset1.rename(columns={"PRODUTC": "PRODUCT"}, inplace=True)
df_dataset1

PRODUCT  VALUE               DATETIME  ... COUNTRY_B     CELL_ID     SITE_ID
0      Voice      0  2012-05-06 23:04:37.0  ...     19e2e  /TJNe+Mmtu  /TJNe+Mmtu
1        sms      0  2012-05-06 23:05:25.0  ...     19e2e  /+cKZKIp41  /+cKZKIp41
2        sms     25  2012-05-06 23:05:41.0  ...     19e2e  /u0FSD+ahi  /u0FSD+ahi
3      Voice     56  2012-05-06 23:05:42.0  ...     19e2e  +cNeJzsTp3  +cNeJzsTp3
4      Voice     14  2012-05-06 23:06:29.0  ...     19e2e  051KhYDCpv  051KhYDCpv
...      ...    ...                    ...  ...       ...         ...         ...
4996     sms    100  2012-05-07 00:01:54.0  ...     OTHER  0c2371d9bc  1e751d08a4
4997   Voice      1  2012-05-07 00:01:54.0  ...     19e2e  5fb791cf56  65efe2bc62
4998   Voice      0  2012-05-07 00:01:54.0  ...     19e2e  7773627c12  9ef9616350
4999   Voice      0  2012-05-07 00:01:54.0  ...     19e2e  3f4af49d6c  969b8d0e39
5000   Voice      7  2012-05-07 00:01:54.0  ...     19e2e  6f4d9b77a5  ebfc58af70

[5001 rows x 10 columns]

renaming product colunm by spelling the name right

In [250]:
#3a.load dataset2
#
df_dataset2= pd.read_csv("dataset2.csv")
df_dataset2

PRODUCT  VALUE DATE_TIME  ... COUNTRY_B     CELL_ID     SITE_ID
0        sms      0   02:06.0  ...     19e2e  +854AcBQT2  +854AcBQT2
1      Voice      0   03:44.0  ...     19e2e  +854AcBQT2  +854AcBQT2
2        sms      0   04:06.0  ...     19e2e      #NAME?      #NAME?
3      Voice     37   04:48.0  ...     19e2e  /xvRrCVKoQ  /xvRrCVKoQ
4        sms     25   04:58.0  ...     19e2e      #NAME?      #NAME?
...      ...    ...       ...  ...       ...         ...         ...
4996     sms      0   01:53.0  ...     19e2e  896459aace  b7d00e59e4
4997     sms     25   01:53.0  ...     19e2e  7e90b35166  211015b14d
4998     sms      0   01:53.0  ...     19e2e  fe70e431df  c7ea989264
4999    data      0   01:53.0  ...     OTHER  ffa6759bb2         NaN
5000   Voice      0   01:53.0  ...     19e2e  8630058aca  b47a998a3e

[5001 rows x 10 columns]

In dataset2 dataset, we had 5001 cells.



In [251]:
#3b.renaming colunm for dataset 2
#
df_dataset2.rename(columns={"DATE_TIME":"DATETIME", "DW_A_NUMBER": "DW_A_NUMBER_INT", "DW_B_NUMBER":"DW_B_NUMBER_INT"}, inplace=True)
df_dataset2.head()

PRODUCT  VALUE DATETIME  ... COUNTRY_B     CELL_ID     SITE_ID
0     sms      0  02:06.0  ...     19e2e  +854AcBQT2  +854AcBQT2
1   Voice      0  03:44.0  ...     19e2e  +854AcBQT2  +854AcBQT2
2     sms      0  04:06.0  ...     19e2e      #NAME?      #NAME?
3   Voice     37  04:48.0  ...     19e2e  /xvRrCVKoQ  /xvRrCVKoQ
4     sms     25  04:58.0  ...     19e2e      #NAME?      #NAME?

[5 rows x 10 columns]

In [252]:
#4a.load dataset3
#
df_dataset3= pd.read_csv("dataset3.csv")
df_dataset3

PRODUCT  VALUE DATE_TIME  ... COUNTRY_B      CELLID     SIET_ID
0      Voice     61   01:28.0  ...     19e2e  +38u2u/rfx  +38u2u/rfx
1        sms      0   01:31.0  ...     19e2e      #NAME?      #NAME?
2      Voice     10   02:59.0  ...     19e2e  +ZaaZozy/+  +ZaaZozy/+
3        sms      0   03:04.0  ...     19e2e      #NAME?      #NAME?
4      Voice     22   03:16.0  ...     19e2e  /eeMklgyJA  /eeMklgyJA
...      ...    ...       ...  ...       ...         ...         ...
4996   Voice      0   01:43.0  ...     19e2e  2aa04dcab7  f7d63a29e5
4997   Voice     62   01:43.0  ...     19e2e  7f1f2f889f  b4ddb9e93c
4998   Voice     32   01:43.0  ...     19e2e  a4731e780e  d68260ef50
4999   Voice     30   01:43.0  ...     19e2e  98ed726da2  b4130b54ab
5000   Voice     60   01:43.0  ...     19e2e  1c11a2fcbb  6d70dae094

[5001 rows x 10 columns]

In dataset3 dataset, we had 5001 cells.

In [253]:
#4b. renaming colunm for dataset3
#
df_dataset3.rename(columns={"DATE_TIME":"DATETIME", "CELLID": "CELL_ID", "ITE_ID":"SITE_ID"}, inplace=True)
df_dataset3.head()

PRODUCT  VALUE DATETIME  ... COUNTRY_B     CELL_ID     SIET_ID
0   Voice     61  01:28.0  ...     19e2e  +38u2u/rfx  +38u2u/rfx
1     sms      0  01:31.0  ...     19e2e      #NAME?      #NAME?
2   Voice     10  02:59.0  ...     19e2e  +ZaaZozy/+  +ZaaZozy/+
3     sms      0  03:04.0  ...     19e2e      #NAME?      #NAME?
4   Voice     22  03:16.0  ...     19e2e  /eeMklgyJA  /eeMklgyJA

[5 rows x 10 columns]

renamed 3 colunms namely date time, cell id and site id.

In [254]:
#5.merging the 3 datasets.
#5a. merging df_dataset1 and df_dataset2 and dataset3
#use concat since all the colunm namses are similer.
df4_telecom= pd.concat([df_dataset1, df_dataset2, df_dataset3], axis =0)
df4_telecom

PRODUCT  VALUE               DATETIME  ...     CELL_ID     SITE_ID     SIET_ID
0      Voice      0  2012-05-06 23:04:37.0  ...  /TJNe+Mmtu  /TJNe+Mmtu         NaN
1        sms      0  2012-05-06 23:05:25.0  ...  /+cKZKIp41  /+cKZKIp41         NaN
2        sms     25  2012-05-06 23:05:41.0  ...  /u0FSD+ahi  /u0FSD+ahi         NaN
3      Voice     56  2012-05-06 23:05:42.0  ...  +cNeJzsTp3  +cNeJzsTp3         NaN
4      Voice     14  2012-05-06 23:06:29.0  ...  051KhYDCpv  051KhYDCpv         NaN
...      ...    ...                    ...  ...         ...         ...         ...
4996   Voice      0                01:43.0  ...  2aa04dcab7         NaN  f7d63a29e5
4997   Voice     62                01:43.0  ...  7f1f2f889f         NaN  b4ddb9e93c
4998   Voice     32                01:43.0  ...  a4731e780e         NaN  d68260ef50
4999   Voice     30                01:43.0  ...  98ed726da2         NaN  b4130b54ab
5000   Voice     60                01:43.0  ...  1c11a2fcbb         NaN  6d70dae094

[15003 rows x 11 columns]

merged 3 tables(df_dataset1, df_dataset2, and dataset3) together. the merged table had 15003 cells.

In [255]:
#5b.summary statistics
#grouping value by product
df4_telecom.groupby(['PRODUCT'])['VALUE'].describe()

count       mean         std  min  25%   50%   75%      max
PRODUCT                                                              
Voice    5666.0  55.169961  236.315367  0.0  1.0  20.0  53.0  12900.0
data     2266.0   7.628861  104.899256  0.0  0.0   0.0   0.0   4440.0
sms      7071.0   7.555791   14.680076  0.0  0.0   0.0  25.0    100.0

merged table has 5666 voice, 2266 data and 7071 sms based on the product.

there was a maximum of 12900 voice, 4440 of data, and 100 sms.

the mean was 55.1 voice, 7.6 data, and 14.6 sms.

In [256]:
#cheking by count
#
df4_telecom.groupby(['PRODUCT'])['VALUE'].count()

PRODUCT
Voice    5666
data     2266
sms      7071
Name: VALUE, dtype: int64

In [266]:
#6. merging df4_telecom and cells_geo dataset
#
#use merge function.
#
df5_telecom= df4_telecom.merge(df_cell_geo_2, left_on="SITE_ID", right_on ="SITE_CODE")
df5_telecom


PRODUCT  VALUE  ...   CELL_ID_y   SITE_CODE
0         sms      0  ...  24ccdda3d5  1b5540c02d
1         sms      0  ...  ebfc58af70  1b5540c02d
2         sms      0  ...  ba8ce6e4ef  1b5540c02d
3         sms      0  ...  c2545171dc  1b5540c02d
4         sms      0  ...  f15bd5483f  1b5540c02d
...       ...    ...  ...         ...         ...
31798   Voice      0  ...  28e79c6e8c  c58809bff2
31799   Voice      0  ...  d7f25faa49  c58809bff2
31800   Voice    180  ...  e1ffe01ec9  27529179de
31801   Voice    180  ...  be4d66264b  27529179de
31802   Voice    180  ...  1099e0473d  27529179de

[31803 rows x 22 columns]

the final merged dataset had 32407 cells.

In [328]:
# summary statistics
#7.Which ones were the most used city for the three days
#
df5_telecom['PRODUCT'].groupby(df5_telecom['VILLES']).describe()


count unique    top  freq
VILLES                                 
ABENGOUROU      111      2  Voice    69
ABOBO          3710      2    sms  2383
ABOISSO          42      2  Voice    24
ABOISSO COMOE    15      2  Voice    12
ABONGOUA          3      1  Voice     3
...             ...    ...    ...   ...
ZATTA             3      1  Voice     3
ZIKISSO          28      2    sms    20
ZON-HOUNIEN       9      1  Voice     9
ZOUKOUGBEU       12      2    sms     6
ZUENOULA        108      2  Voice    63

[373 rows x 4 columns]

ABOBO was the highest product usage city.

In [324]:
#
#8.Which cities were the most used during business and home hours
#
df5_telecom['DATETIME'].groupby(df5_telecom['VILLES']).describe()


count unique                    top freq
VILLES                                                
ABENGOUROU      111     21  2012-05-07 00:01:22.0   12
ABOBO          3710    220  2012-05-07 00:01:26.0   47
ABOISSO          42      7  2012-05-07 00:00:41.0    6
ABOISSO COMOE    15      5                00:48.0    3
ABONGOUA          3      1                00:26.0    3
...             ...    ...                    ...  ...
ZATTA             3      1  2012-05-07 00:01:53.0    3
ZIKISSO          28      7                01:53.0    4
ZON-HOUNIEN       9      3                01:35.0    3
ZOUKOUGBEU       12      4  2012-05-07 00:00:48.0    3
ZUENOULA        108     34                00:34.0    6

[373 rows x 4 columns]

ABOBO was the most used city during business and home hours

> Indented block



In [327]:
#check product by localisation
df5_telecom['PRODUCT'].groupby(df5_telecom['LOCALISATION']).describe()


count unique    top   freq
LOCALISATION                            
ABIDJAN       19334      2    sms  11105
INTERIEUR     12469      2  Voice   6980

ABIJAN was the highest in product usage based on **localisation** followed by INTERIEUR

In [329]:
#check product usage based on region
df5_telecom['PRODUCT'].groupby(df5_telecom['REGION']).describe()

count unique    top   freq
REGION                                       
18 MONTAGNES         630      2  Voice    327
AGNEBY               405      2  Voice    216
BAFING                99      2  Voice     63
BAS SASSANDRA       1500      2    sms    785
DENGUELE             320      2  Voice    206
FROMAGER             566      2  Voice    306
HAUT SASSANDRA      1655      2  Voice    833
LACS                1449      2  Voice    728
LAGUNE             20163      2    sms  11580
MARAHOUE             477      2  Voice    291
MOYEN CAVALLY        450      2  Voice    273
MOYEN COMOE          219      2  Voice    153
NZI COMOE            370      2  Voice    235
SAVANE               661      2  Voice    473
SUD BANDAMA          730      2  Voice    410
SUD COMOE            123      2  Voice     81
VALLEE DU BANDAMA    981      2  Voice    620
WORODOUGOU           696      2  Voice    480
ZANZAN               309      2  Voice    216

lagune was the highest in the region.